src/core/pointcloud/qgscopcpointcloudindex.cpp

```diff
@@ -181,6 +186,68 @@ bool QgsCopcPointCloudIndex::loadHierarchy()
   return true;
 }
 
+bool QgsCopcPointCloudIndex::writeStatistics( QgsPointCloudStatistics &stats )
+{
+  if ( mLazInfo->version() != qMakePair<uint8_t, uint8_t>( 1, 4 ) )
+  {
+    // EVLR isn't supported in the first place
+    QgsMessageLog::logMessage( tr( "Can't write statistics to \"%1\": laz version != 1.4" ).arg( mFileName ) );
+    return false;
+  }
+
+  QByteArray statisticsEvlrData = fetchCopcStatisticsEvlrData();
+  if ( !statisticsEvlrData.isEmpty() )
+  {
+    QgsMessageLog::logMessage( tr( "Can't write statistics to \"%1\": file already contains COPC statistics!" ).arg( mFileName ) );
+    return false;
+  }
+
+  lazperf::evlr_header statsEvlrHeader;
+  statsEvlrHeader.user_id = "qgis";
+  statsEvlrHeader.record_id = 0;
+  statsEvlrHeader.description = "Contains calculated statistics";
+  QByteArray statsJson = stats.toStatisticsJson();
+  statsEvlrHeader.data_length = statsJson.size();
+
+  // Save the EVLRs to the end of the original file (while erasing the exisitng EVLRs in the file)
+  mCopcFile.close();
+  std::fstream copcFile;
+  copcFile.open( QgsLazDecoder::toNativePath( mFileName ), std::ios_base::binary | std::iostream::in | std::iostream::out );
+  if ( copcFile.is_open() && copcFile.good() )
+  {
+    // Write the new number of EVLRs
+    lazperf::header14 header = mLazInfo->header();
+    header.evlr_count = header.evlr_count + 1;
+    copcFile.seekp( 0 );
+    header.write( copcFile );
+
+    // Append EVLR data to the end
+    copcFile.seekg( 0, std::ios::end );
+
+    statsEvlrHeader.write( copcFile );
+    copcFile.write( statsJson.data(), statsEvlrHeader.data_length );
+  }
+  else
+  {
+    QgsMessageLog::logMessage( tr( "Couldn't open COPC file \"%1\" to write statistics" ).arg( mFileName ) );
+    return false;
+  }
+  mCopcFile.open( QgsLazDecoder::toNativePath( mFileName ), std::ios::binary );
```

Comment:
At the time we re-open `mCopcFile`, the `copcFile` fstream is still opened. I think it would be safer to explicitly close `copcFile` before re-opening it, to make sure there are no pending writes...